### PR TITLE
feat: JWT to authenticate external CL against Reth

### LINKED ISSUE #1098: JWT to authenticate external CL against Reth
### Describe the feature

I propose Reth to adhere the JWT standard to setup a trust relationship with an external CL.

Given the current code base and available libraries I'd suggest to create a `tower::Layer` to cope with JWT authentication.
At a first glance the code could look like this:

```rust
   // Read these args from CLI
   // --authrpc.addr=0.0.0.0
   // --authrpc.port=8551
   // --authrpc.jwtsecret=/etc/jwt/jwtsecret

    // Read these from CLI
    const AUTH_JWT: &str = "/etc/jwt/jwtsecret"; 
    const AUTH_PORT: u32 = 8551; 
    const AUTH_ADDR: &str = "0.0.0.0";

    // RPC layer communicates with the implemented logic
    // through an unbounded channel
    let (tx, rx) = unbounded_channel();
   
    let chain_spec = MAINNET.clone();
    let client = Arc::new(MockEthProvider::default()); // Here we should provide a real Client implementation

    // EthConsensusEngine is already implemented as an endless future
    // that listens on the rx part of the channel. 
    // Let's spawn a task.
    tokio::spawn(EthConsensusEngine {
        client: client.clone(),
        chain_spec: chain_spec.clone(),
        local_store: Default::default(),
        rx: UnboundedReceiverStream::new(rx),
    });

    // We need to implement JwtLayer as a `tower::Layer`.
    // This layer checks a valid JWT is provided in CL HTTP requests
    let jwt_layer = JwtLayer::new(AUTH_JWT)?; // Invalid JWT make this fail
    let jwt_middleware = tower::ServiceBuilder::default().layer(jwt_layer);
    
    let server = ServerBuilder::default()
        .set_middleware(jwt_middleware)
        .build(format!("{AUTH_ADDR}:{AUTH_PORT}").parse::<SocketAddr>()?)
        .await?;

    let rpc = EngineApi { engine_tx: tx }.into_rpc();
    let handle = server.start(rpc)?;
```

What do you guys think?

### Additional context

I suppose all this stuff applies if Reth is built with `feature = "client"`

### PR DESCRIPTION
**Pull request by 0xMelkor**
Closes #1098

### CODE REVIEW

File: crates/net/rpc/src/engine/jwt_secret.rs [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+/// [Secret key - Engine API specs](https://github.com/ethereum/execution-apis/blob/main/src/engine/authentication.md#key-distribution)
+#[derive(Debug)]
+pub struct JwtSecret(String);
+
+impl JwtSecret {
+    /// Creates an instance of [`JwtSecret`][crate::engine::JwtSecret].
+    /// Generates and error if one of the following applies:
+    /// - The `hex` argument is not a valid hexadecimal string
+    /// - The `hex` argument length is less than `JWT_SECRET_LEN`
+    pub fn new(hex: String) -> Result<JwtSecret, JwtError> {
```
Comment (mattsse) [inline crates/net/rpc/src/engine/jwt_secret.rs:55]: this is essentially a `from_hex` function, right?

suggest renaming

**Review by mattsse — CHANGES_REQUESTED** [review]
this is a great draft, left some suggestions and nits.

next step would be an Auth Layer implementation, see for example:

https://docs.rs/tower-http/latest/src/tower_http/auth/require_authorization.rs.html

**Comment by 0xMelkor** [comment]
@mattsse, thank you for reviewing this draft.
I'm thinking to a clean approach to implement the JWT layer. I'll consider also your link.

File: crates/net/rpc/src/engine/jwt_secret.rs (hunk 1 above)
Comment (gakonst) [reply crates/net/rpc/src/engine/jwt_secret.rs:55]: Or just `TryFrom<String>` and make it `JwtSecret(Bytes)` or something.

File: crates/net/rpc/src/engine/jwt_secret.rs [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+    pub fn random() -> Self {
+        let random_bytes: [u8; 32] = rand::thread_rng().gen();
+        let secret = hex_encode(random_bytes);
+        Self(secret)
+    }
+
+    /// Returns a string representation of the secret
+    pub fn as_string(&self) -> String {
+        self.0.clone()
+    }
```
[the remark this answers is not in the document]
Comment (gakonst) [reply crates/net/rpc/src/engine/jwt_secret.rs:75]: remove and make the Secret be over Bytes as written above

**Comment by gakonst** [comment]
@mattsse re: "next step would be an Auth Layer implementation", we don't use tower outside of IPC, so what would be the right way to impl that?

also, do we have the code necessary for spawning engine api on a separate port?

**Review by gakonst — COMMENTED** [review]
love the direction - i've rebased the PR for conflicts

**Comment by 0xMelkor** [comment]
Pushed a brief version of the Jwt layer (need to add some testing). 

I've some points:
1. Should we return an error body along with the `401` status? Any specs out there?
2. Should we provide a trait for the auth header verifier, to allow middleware reuse?

**Comment by mattsse** [comment]
> @mattsse re: "next step would be an Auth Layer implementation", we don't use tower outside of IPC, so what would be the right way to impl that?

we do, all rpc server support tower middleware

> also, do we have the code necessary for spawning engine api on a separate port?

this should be part of the rpc-builder eventually

File: crates/rpc/rpc/src/engine/jwt_layer.rs [hunk 3]
```diff
@@ -0,0 +1,134 @@
+use http::{HeaderMap, Request, Response, StatusCode};
+
```
Comment (mattsse) [inline crates/rpc/rpc/src/engine/jwt_layer.rs:2]:
```suggestion
```

File: crates/rpc/rpc/src/engine/jwt_layer.rs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+impl<S> Layer<S> for JwtLayer {
+    type Service = JwtService<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        JwtService { secret: self.0.clone(), inner }
+    }
+}
+
+#[allow(missing_debug_implementations)]
```
Comment (mattsse) [inline crates/rpc/rpc/src/engine/jwt_layer.rs:33]: needs docs

File: crates/rpc/rpc/src/engine/jwt_layer.rs [hunk 5]
```diff
@@ -0,0 +1,134 @@
+use http::{HeaderMap, Request, Response, StatusCode};
+
+use pin_project::pin_project;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower::{Layer, Service};
+
```
Comment (mattsse) [inline crates/rpc/rpc/src/engine/jwt_layer.rs:10]:
```suggestion
```

File: crates/rpc/rpc/src/engine/jwt_layer.rs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+impl<S> JwtService<S> {
+    // TODO: Add full coverage
+    fn get_bearer(&self, headers: &HeaderMap) -> Option<String> {
+        headers
+            .get("authorization")
+            .map(|header| header.to_str())
+            .and_then(|result| result.ok())
+            .and_then(|auth| {
```
Comment (mattsse) [inline crates/rpc/rpc/src/engine/jwt_layer.rs:76]: can use `?` instead

File: crates/rpc/rpc/src/engine/jwt_layer.rs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+                let prefix = "Bearer ";
+                auth.find(prefix)
+                    .map(|index| &auth[index + prefix.len()..])
+                    .map(|token| token.to_owned())
```
Comment (mattsse) [inline crates/rpc/rpc/src/engine/jwt_layer.rs:80]: this looks a bit weird, perhaps can use 
`Bearer` type https://docs.rs/tower-http/latest/src/tower_http/auth/require_authorization.rs.html#382
or mirror how this type does auth, basically tracking the expected HeaderValue

File: crates/rpc/rpc/src/engine/jwt_secret.rs [hunk 8]
```diff
@@ -0,0 +1,236 @@
+use std::time::{SystemTime, UNIX_EPOCH};
+
```
Comment (mattsse) [inline crates/rpc/rpc/src/engine/jwt_secret.rs:2]:
```suggestion
```

File: crates/rpc/rpc/src/engine/jwt_secret.rs [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+/// Length of the hex-encoded 256 bit secret key.
+/// A 256-bit encoded string in Rust has a length of 64 digits because each digit represents 4 bits
+/// of data. In hexadecimal representation, each digit can have 16 possible values (0-9 and A-F), so
+/// 4 bits can be represented using a single hex digit. Therefore, to represent a 256-bit string,
+/// we need 64 hexadecimal digits (256 bits ÷ 4 bits per digit = 64 digits).
+const JWT_SECRET_LEN: usize = 64;
+
+/// The JWT `iat` (issued-at) claim cannot exceed +-60 seconds from the current time.
+const JWT_MAX_IAT_DIFF: u64 = 60;
```
Comment (mattsse) [inline crates/rpc/rpc/src/engine/jwt_secret.rs:35]: please use Duration

File: crates/rpc/rpc/src/engine/jwt_secret.rs [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+const JWT_SIGNATURE_ALGO: Algorithm = Algorithm::HS256;
+
+/// Value-object holding a reference to an hex-encoded 256 bit secret key.
+/// A JWT secret key is used to secure JWT-based authentication. The secret key is
+/// a shared secret between the server and the clienta and is used to calculate a digital signature
+/// for the JWT, which is included in the JWT along with its payload.
+/// -----------------------------------------------------------------
+/// [Secret key - Engine API specs](https://github.com/ethereum/execution-apis/blob/main/src/engine/authentication.md#key-distribution)
+#[derive(Clone)]
+#[allow(missing_debug_implementations)]
```
Comment (mattsse) [inline crates/rpc/rpc/src/engine/jwt_secret.rs:47]:
```suggestion
```

**Review by mattsse — CHANGES_REQUESTED** [review]
good progress.

left some suggestions.

we need a example/test for this,
see https://github.com/paritytech/jsonrpsee/blob/master/examples/examples/middleware.rs

File: crates/net/rpc/src/engine/jwt_secret.rs (hunk 1 above)
Comment (0xMelkor) [reply crates/net/rpc/src/engine/jwt_secret.rs]: I implemented the TryInto trait, it looks nicer.
I also builds `JwtSecret` on top fo bytes

File: crates/rpc/rpc/src/engine/jwt_secret.rs (hunk 10 above)
Comment (0xMelkor) [reply crates/rpc/rpc/src/engine/jwt_secret.rs:47]: I'd disallow printing the JWT secret to the console. 
I could provide a `Display` impl that returns an hash or some obfuscated value.

File: crates/rpc/rpc/src/engine/jwt_secret.rs [hunk 11]
```diff
@@ -45,23 +45,38 @@ const JWT_SIGNATURE_ALGO: Algorithm = Algorithm::HS256;
 /// [Secret key - Engine API specs](https://github.com/ethereum/execution-apis/blob/main/src/engine/authentication.md#key-distribution)
 #[derive(Clone)]
 #[allow(missing_debug_implementations)]
-pub struct JwtSecret(String);
+pub struct JwtSecret([u8; 32]);
+
+impl TryInto<JwtSecret> for String {
```
Comment (mattsse) [inline crates/rpc/rpc/src/engine/jwt_secret.rs:54]: I'm not a fan of this `TryInto`, I'd rather have a `from_hex(AsRef<str>)`, which is more explicit.

File: crates/rpc/rpc/src/engine/jwt_secret.rs [hunk 12]
```diff
@@ -45,23 +45,38 @@ const JWT_SIGNATURE_ALGO: Algorithm = Algorithm::HS256;
...
         if hex.len() != JWT_SECRET_LEN {
             Err(JwtError::InvalidLength(JWT_SECRET_LEN, hex.len()))
         } else {
-            hex::decode(&hex)?;
-            Ok(Self(hex))
+            let hex_bytes = hex::decode(&hex)?;
+            let mut bytes = [0; 32];
+            let mut writer = &mut bytes[..];
+            let _ = writer.write(hex_bytes.get(0..32).expect("This should never happen"));
+            Ok(JwtSecret(bytes))
```
Comment (mattsse) [inline crates/rpc/rpc/src/engine/jwt_secret.rs:70]: can use `bytes.as_slice().try_into()` here directly

File: crates/rpc/rpc/src/engine/jwt_layer.rs (hunk 4 above)
Comment (0xMelkor) [reply crates/rpc/rpc/src/engine/jwt_layer.rs:87]: Docs added, tell me if ok. Thanks sir.

File: crates/rpc/rpc/src/engine/jwt_secret.rs (hunk 10 above)
Comment (0xMelkor) [reply crates/rpc/rpc/src/engine/jwt_secret.rs]: Added a `Display` impl that computes an hash with standard Rust facilities.

File: crates/rpc/rpc/src/engine/jwt_layer.rs [hunk 13]
```diff
@@ -30,9 +80,15 @@ impl<S> Layer<S> for JwtLayer {
     }
 }
 
+/// This trait is the actual implementation of
```
Comment (mattsse) [inline crates/rpc/rpc/src/engine/jwt_layer.rs:83]:
```suggestion
/// This type is the actual implementation of
```

File: crates/rpc/rpc/src/engine/jwt_secret.rs (hunk 12 above)
Comment (0xMelkor) [reply crates/rpc/rpc/src/engine/jwt_secret.rs:68]: You refer to the instantiation of `JwtSecret`? If so I removed the `TryInto` impl, so it's implicitly solved.
Otherwise can you be more clear?

File: crates/stages/src/stages/hashing_account.rs [hunk 14]
```diff
@@ -15,7 +15,7 @@ use std::{
 use tracing::*;
 
 /// The [`StageId`] of the account hashing stage.
-pub const ACCOUNT_HASHING: StageId = StageId("AccountHashing");
+pub const ACCOUNT_HASHING: StageId = StageId("AccountHashingStage");
```
Comment (onbjerg) [inline crates/stages/src/stages/hashing_account.rs:18]: This is unrelated, please remove it from the PR

Comment (0xMelkor) [reply crates/stages/src/stages/hashing_account.rs:18]: This is a side effect of a rebase. I'll remove it

**Comment by 0xMelkor** [comment]
@mattsse @onbjerg I need some directions:
The current layer returns a `401` status code with an empty error payload in case of bad requests. According to [RPC specs](https://www.jsonrpc.org/specification#error_object) this is correct.
Anyway I don't know if this is the intended behavior for the "Engine API" (I didn't find specs out there)

Should we provide a JSON-RPC error payload instead? (I've some design ideas on that)

My opinion is that authorization is a transport logic concern, that must be kept separate from application errors.

What's your opinion?

File: crates/rpc/rpc/src/engine/jwt_layer.rs (hunk 7 above)
Comment (0xMelkor) [reply crates/rpc/rpc/src/engine/jwt_layer.rs]: > Bearer type https://docs.rs/tower-http/latest/src/tower_http/auth/require_authorization.rs.html#382

The code in the link makes a comparison like `"Bearer <ACTUAL>" == "Bearer <EXPECTED>"`.
We have a different requirement: extract the token value after the "Bearer" prefix to perform a deeper validation.

I simplified the code, tell me if you like it.

**Comment by gakonst** [comment]
(@0xMelkor would look at how Erigon/Geth does it)

File: crates/rpc/rpc/src/layers/auth_layer.rs [hunk 15]
```diff
@@ -0,0 +1,291 @@
+use http::{Request, Response};
+use http_body::Body;
+use pin_project::pin_project;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower::{Layer, Service};
+
```
Comment (mattsse) [inline crates/rpc/rpc/src/layers/auth_layer.rs:10]:
```suggestion
```

File: crates/rpc/rpc/src/layers/jwt_validator.rs [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+                Err(response)
+            }
+        }
+    }
+}
+
+/// This is an utility function that retrieves a bearer
+/// token from an authorization Http header.
+fn get_bearer(headers: &HeaderMap) -> Option<String> {
+    let header = headers.get("authorization")?;
```
Comment (mattsse) [inline crates/rpc/rpc/src/layers/jwt_validator.rs:49]: please the `AUTHORIZATION` Headervalue constant.

**Review by mattsse — CHANGES_REQUESTED** [review]
smol nits

File: crates/rpc/rpc/src/layers/auth_layer.rs (hunk 15 above)
Comment (0xMelkor) [reply crates/rpc/rpc/src/layers/auth_layer.rs:10]: `cargo +nightly fmt` rules on this.

**Comment by 0xMelkor** [comment]
Hi guys,
I think this is quite mature now. If review is ok, I would iterate and open new issues for simpler tracking:

1. **JWT secret lifecycle**: 
  . Get path from CLI then load from disk
  . Create ephemeral secret if not provided by user then store on disk
2. **Extend the Engine API**: Add recommended `eth_` namespace methods for the Engine API
3. **Virtual hosts support**: Provide a new middleware to filter virtual hosts on incoming calls(`--authrpc.vhosts`)
4. **Spawn the engine server**
ecc.

Of course I would be happy to help

**Review by mattsse — APPROVED** [review]
lgtm
ty.
And yes please continue with the follow-ups you listed.
